File: amazing_maze_scala/src/main/scala/enums/Maze.scala
package enums

import enums.Exploration.*

import scala.::
import scala.collection.mutable.ListBuffer

enum Maze:
  case Branch(label: String, left: Maze, right: Maze, var status: Exploration = UnExplored)
  case Leaf(label: String)

  def explore_old: List[String] = this match
    case branch@Branch(label, left, right, status) =>
      status match
        case UnExplored => branch.status = Explored; label :: left.explore_old ::: right.explore_old
        case Explored => List(label)
    case Leaf(label) => List(label)

  def explore(trace: ListBuffer[String]): Unit =
    this match
      case branch@Branch(label, left, right, status) =>
        status match
          case UnExplored => branch.status = Explored
            trace += label
            left.explore(trace)
            right.explore(trace)
          case Explored => trace += label
      case Leaf(label) => trace += label

File: amazing_maze_rust/src/enums/maze.rs
// 1. Définir la structure d'un labyrinthe et d'implémenter l'exemple

use std::cell::RefCell;

use Exploration::*;

use crate::enums::maze::Maze::Branch;
use crate::enums::maze::Maze::Leaf;

use super::exploration::Exploration;

pub enum Maze<'a> {
    Branch(String, &'a Maze<'a>, &'a Maze<'a>, RefCell<Exploration>),
    Leaf(String),
}

impl Maze<'_> {
    pub fn explore(&self, buffer: &mut Vec<String>) {
        match self {
            Branch(label, left, right, statut) => {
                if *statut.borrow() == UnExplored {
                    statut.replace(Explored);
                    buffer.push(label.to_string());
                    left.explore(buffer);
                    right.explore(buffer);
                } else {
                    buffer.push(label.to_string());
                }
            }
            Leaf(label) => buffer.push(label.to_string()),
        }
    }
}
